src/index.ts

```diff
@@ -12,20 +12,24 @@ export class Iterable<T> implements LibIterable<T> {
 
     /**
      * Creates an instance of Iterable.
-     * @param {IterableIterator<T>} source The source iterable iterator to use. This can be any object that supports the {Symbol.iterator} symbol.
+     * @param {Iterable<T>} source The source iterable to use. This can be any object that supports
+     * the {Symbol.iterator} symbol, or a function that returns an Iterable (e.g., a generator)
      * @memberof Iterable
      */
-    constructor(source: { [Symbol.iterator]: () => Iterator<T> }) {
-        if (!source) {
-            throw new ReferenceError(`Invalid source. source is '${source}'`);
-        }
-        if (!source[Symbol.iterator]) {
+    constructor(source: LibIterable<T> | (() => LibIterable<T>)) {
+        if (typeof source === 'function') {
+            // source is an function that returns an Iterable. Get its iterator lazily:
+            this._source = {
```

Comment:
I'd ideally like to move all this logic to the iterator function to avoid creating a dummy object:
```ts
[Symbol.iterator]() {
    return (typeof this.source === 'function' ? this.source() : this.source)[Symbol.iterator]();
}
```

and change all the utility functions to:
```ts
for (const item of this) { ... }
```

This way, we use the iterator function exactly like how external consumers would use it.